Support CREATE/DROP FULLTEXT INDEX for fulltext search

roshan0708 opened the pull request:
### Description

Currently babelfish doesn't support fulltext index creation and drop for full text search. This commit implements the logic of create and drop fulltext index and also handles the unsupported features of the CREATE/DROP FULLTEXT INDEX statements. 

Signed-off-by: Roshan Kanwar [rskanwar@amazon.com](mailto:rskanwar@amazon.com)

### Issues Resolved

Task: BABEL-4383

### Test Scenarios Covered ###
* **Use case based -** Yes


* **Boundary conditions -** Yes


* **Arbitrary inputs -**


* **Negative test cases -** Yes


* **Minor version upgrade tests -**


* **Major version upgrade tests -**


* **Performance tests -**


* **Tooling impact -**


* **Client tests -**



### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

roshan0708 committed:
Support CREATE/DROP FULLTEXT INDEX for fulltext search

Currently babelfish doesn't support fulltext index creation and drop for
fulltext search. This commit implements the logic of create and drop
fulltext index and also handles the unsupported features of the
CREATE/DROP FULLTEXT INDEX statements.

Task: BABEL-4383
Signed-off-by: Roshan Kanwar <rskanwar@amazon.com>

roshan0708 committed:
Merge remote-tracking branch 'upstream/BABEL_3_X_DEV' into JIRA-BABEL-4383

surendravishnoi commented on contrib/babelfishpg_tsql/src/guc.c:
```diff
@@ -1336,13 +1336,13 @@ define_escape_hatch_variables(void)
 							 escape_hatch_options,
-							 PGC_USERSET,
-							 GUC_NOT_IN_SAMPLE | GUC_DISALLOW_IN_FILE | GUC_DISALLOW_IN_AUTO_FILE,
+							 PGC_SUSET,
```
what is the purpose of this change?

roshan0708 replied:
This is to only allow superuser to enable full text search, initially we were using escape hatch which the customer can use to enable fts anytime which we don’t want. After pentest, if everything goes right, we will remove this.

surendravishnoi replied:
escape hatch usually means that when set corresponding commands will be ignored so that application can work. What will we do in case of FTS feature rollback? how this change will help?

roshan0708 replied:
If a normal user tries to use this escape hatch to turn on fulltext search, it will throw `permission error`. This can only be turned on if the login is of super user. We have just put it like this so that let's say we roll FTS feature in prod, the customer shouldn't be able to use it rather only we, as superuser, should be able to turn it on. We'll rollback these extra changes of escape hatch once the pentest is done, that's what we have planned so that the customer is able to use it normally without turning on any escape hatch.

surendravishnoi commented on contrib/babelfishpg_tsql/src/pl_exec-2.c:
```diff
@@ -3658,6 +3661,16 @@ get_insert_bulk_kilobytes_per_batch()
 }
 
+static const char 
+*gen_schema_name_for_fulltext_index(char *schema_name)
```
should be check for schema name being NULL?

roshan0708 replied:
That is already handled in tsqlIface file, while I’m extracting other data items.

Sairakan replied:
Just for sanity, I think schema_name should be declared as const here, e.g. `const char *schema_name`

roshan0708 replied:
In the query statement, if we have not specified the schema name, it will be an empty string i.e. `""` else we get the proper schema name in the helper function. So the schema name will never be null. Here is the below code for reference - 
```
if (ctx->table_name())
{
	std::string table_info = ::getFullText(ctx->table_name());
	std::string table_name = "";
	std::string schema_name = "";
	size_t pos = table_info.find(".");
	if (pos != std::string::npos) {
		// Extract the schema name before the "."
		schema_name = table_info.substr(0, pos);
		// Extract the table name after the "."
		table_name = table_info.substr(pos + 1);
	} else {
		// No "." character found, set first to the entire string
		table_name = table_info;
	}
	stmt->table_name = pstrdup(downcase_truncate_identifier(table_name.c_str(), table_name.length(), true));
	stmt->schema_name = pstrdup(downcase_truncate_identifier(schema_name.c_str(), schema_name.length(), true));
}
```

surendravishnoi commented on contrib/babelfishpg_tsql/src/pl_exec-2.c:
```diff
@@ -3817,3 +3830,110 @@ exec_stmt_change_dbowner(PLtsql_execstate *estate, PLtsql_stmt_change_dbowner *s
+	if (!OidIsValid(schemaOid))
+			ereport(ERROR,
+				(errcode(ERRCODE_UNDEFINED_SCHEMA),
+					errmsg("schema \"%s\" does not exist",
```
check for null schema to be on safe side

surendravishnoi commented on contrib/babelfishpg_tsql/src/pl_exec-2.c:
```diff
@@ -3817,3 +3830,110 @@ exec_stmt_change_dbowner(PLtsql_execstate *estate, PLtsql_stmt_change_dbowner *s
+						table_name)));
+
+	// check if fulltext index already exists in the table
+	if (!get_index_name(relid))
```
this can be combined with if else
if (get_index_name(relid)) error else get query str

surendravishnoi commented on contrib/babelfishpg_tsql/src/pl_exec-2.c:
```diff
@@ -3817,3 +3830,110 @@ exec_stmt_change_dbowner(PLtsql_execstate *estate, PLtsql_stmt_change_dbowner *s
+	else
+		ereport(ERROR,
+					(errcode(ERRCODE_WRONG_OBJECT_TYPE),
+					 errmsg("A full-text index for table or indexed view \"%s\" has already been created."
```
also print index name (mostly try to match what error SQL server will give)

surendravishnoi commented on contrib/babelfishpg_tsql/src/pl_exec-2.c:
```diff
@@ -3817,3 +3830,110 @@ exec_stmt_change_dbowner(PLtsql_execstate *estate, PLtsql_stmt_change_dbowner *s
+	bool		index_exists = true;
+
+	table_name = stmt->table_name;
+	schema_name = gen_schema_name_for_fulltext_index((char *)stmt->schema_name);
```
check for null schema

surendravishnoi commented on contrib/babelfishpg_tsql/src/pl_exec-2.c:
```diff
@@ -3817,3 +3830,110 @@ exec_stmt_change_dbowner(PLtsql_execstate *estate, PLtsql_stmt_change_dbowner *s
+}
+
+static int	
+exec_stmt_dropfulltextindex(PLtsql_execstate *estate, PLtsql_stmt_dropfulltextindex *stmt)
```
we can handle both in a single function with a boolean to distinguish create vs drop, most of the code is duplicate

surendravishnoi commented on contrib/babelfishpg_tsql/src/pl_exec-2.c:
```diff
@@ -3817,3 +3830,110 @@ exec_stmt_change_dbowner(PLtsql_execstate *estate, PLtsql_stmt_change_dbowner *s
+					errmsg("relation \"%s\" does not exist",
+						table_name)));
+				
+	index_name = get_index_name(relid);
```
why this assumption that get_index_name will only return full text index? what about other indexes on table? If this function is only for full text index, it should be named accordingly

roshan0708 replied:
This is specifically for fulltext index, will rename it.

surendravishnoi commented on contrib/babelfishpg_tsql/src/pltsql_utils.c:
```diff
@@ -1791,3 +1791,78 @@ List
+		list_free(indexoidlist);
+		return NULL;
+	}
+	indexOid = linitial_oid(indexoidlist);
```
why are we assuming that first index is full text index only?

roshan0708 replied:
Because there can be only one fulltext index per table acc. to SQL Server docs. We get a list of indexes, so we assume we get the fulltext index in the first item only.

surendravishnoi commented on contrib/babelfishpg_tsql/src/pltsql_utils.c:
```diff
@@ -1791,3 +1791,78 @@ List
+}
+
+char
+*gen_createfulltextindex_cmds(const char *table_name, const char *schema_name, const List *column_name, const char *index_name)
```
Please make sure column_name list and index name is not null or we handle null values

roshan0708 replied:
We already handle null values, so if we pass no column or NULL as index name in the query statement, ANTLR will automatically throw `syntax error`.

surendravishnoi commented on contrib/babelfishpg_tsql/src/pltsql_utils.c:
```diff
@@ -1791,3 +1791,78 @@ List
+	 * DROP INDEX <index_name>
+	 *
+	 */
+	appendStringInfo(&query, "DROP INDEX \"%s\".\"%s\"", schema_name, index_name);
```
why no check for schema name being NULL here? also, is it guaranteed that index name is NULL across database?

roshan0708 replied:
Done

surendravishnoi commented on contrib/babelfishpg_tsql/src/tsqlIface.cpp:
```diff
@@ -1802,13 +1820,11 @@ class tsqlBuilder : public tsqlCommonMutator
 			ereport(WARNING,
 					(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
-					 errmsg("FULLTEXT related statements will be ignored.")));
+					 errmsg("ALTER CREATE FULLTEXT INDEX statement will be ignored.")));
```
ALTER FULLTEXT rather than ALTER CREATE FULLINDEX

surendravishnoi commented on contrib/babelfishpg_tsql/src/tsqlIface.cpp:
```diff
@@ -6580,6 +6596,106 @@ post_process_alter_table(TSqlParser::Alter_tableContext *ctx, PLtsql_stmt_execsq
 	return false;
 }
 
+PLtsql_stmt*
```
Indetnantion is messed up

surendravishnoi commented on contrib/babelfishpg_tsql/src/tsqlIface.cpp:
```diff
@@ -6580,6 +6596,106 @@ post_process_alter_table(TSqlParser::Alter_tableContext *ctx, PLtsql_stmt_execsq
+    }
+	if (ctx->catalog_filegroup_option()) {
+		auto catalog_filegroup_option = ctx->catalog_filegroup_option();
+		if (catalog_filegroup_option->FILEGROUP() || catalog_filegroup_option->catalog_name || catalog_filegroup_option->filegroup_name)
```
which filegroup option we support?

roshan0708 replied:
We don't support any filegroup options now.

surendravishnoi replied:
if we do not support any, can we remove if and just give an error for having catalog_filegroup_option

roshan0708 committed:
fixed fts-contains and babelfish configuration
test files for FTS configuration

surendravishnoi commented on contrib/babelfishpg_tsql/src/tsqlIface.cpp:
```diff
@@ -6580,6 +6596,106 @@ post_process_alter_table(TSqlParser::Alter_tableContext *ctx, PLtsql_stmt_execsq
+		if (catalog_filegroup_option->FILEGROUP() || catalog_filegroup_option->catalog_name || catalog_filegroup_option->filegroup_name)
+			throw PGErrorWrapperException(ERROR, ERRCODE_FEATURE_NOT_SUPPORTED, "FILEGROUP/CATALOG option is not supported yet for CREATE FULLTEXT INDEX statement", getLineAndPos(ctx));
+	}
+	if (ctx->fulltext_with_option().size() > 0) {
```
which with options we support?

roshan0708 replied:
We don't support any with options now.

surendravishnoi replied:
same here

surendravishnoi commented on contrib/babelfishpg_tsql/src/tsqlIface.cpp:
```diff
@@ -6580,6 +6596,106 @@ post_process_alter_table(TSqlParser::Alter_tableContext *ctx, PLtsql_stmt_execsq
+	PLtsql_stmt_dropfulltextindex *stmt = (PLtsql_stmt_dropfulltextindex *) palloc0(sizeof(PLtsql_stmt_dropfulltextindex));
+	stmt->cmd_type = PLTSQL_STMT_DROPFULLTEXTINDEX;
+	stmt->lineno = getLineNo(ctx);
+	if (ctx->table_name())
```
getting table name could be a utility function shared between both create/drop

surendravishnoi commented on contrib/babelfishpg_tsql/src/tsqlUnsupportedFeatureHandler.cpp:
```diff
@@ -1057,6 +1057,38 @@ antlrcpp::Any TsqlUnsupportedFeatureHandlerImpl::visitDdl_statement(TSqlParser::
 		if (create_user->WITHOUT())
 			handle(INSTR_UNSUPPORTED_TSQL_UNKNOWN_DDL, "CREATE USER WITHOUT LOGIN",  getLineAndPos(ctx));
 	}
+	if(ctx->create_fulltext_index())
```
why we have same error handling at two places?

Sairakan replied:
+1, if we already are catching unsupported features here, it makes no sense to have the same error handling in the main handler. I would remove the handling in the main function and leave the error handling here.

roshan0708 replied:
Will do

surendravishnoi commented on contrib/babelfishpg_tsql/src/pl_exec-2.c:
```diff
@@ -3817,3 +3830,110 @@ exec_stmt_change_dbowner(PLtsql_execstate *estate, PLtsql_stmt_change_dbowner *s
+
+	// check if table exists
+	if (!OidIsValid(relid))
+			ereport(ERROR,
```
generic comment
1) We should have test case for each error we are throwing, in executor or ANTLR
2) We should have test case for each condition if/else we have. For example : creating full text index on one column vs multiple column

roshan0708 replied:
Yes, I've added those test cases in the JDBC test files of full text index.

roshan0708 committed:
fixed some warnings and other requested changes

Sairakan commented on contrib/babelfishpg_tsql/src/pl_exec-2.c:
```diff
@@ -3662,7 +3661,7 @@ get_insert_bulk_kilobytes_per_batch()
 
 static const char 
-*gen_schema_name_for_fulltext_index(char *schema_name)
+*gen_schema_name_for_fulltext_index(const char *schema_name)
```
I'm not clear on why this function exists separately from other fully qualified schema name resolutions - can you add a comment header to the function explaining why we need this? Is full text search the only case when this kind of fully qualified name is needed?

roshan0708 replied:
Actually, it's just a helper function that I created specifically for getting the schema name for FTS. It's basically just a combination of two functions to handle the master db schema and custom schema case.

Sairakan commented on contrib/babelfishpg_tsql/src/pl_exec-2.c:
```diff
@@ -3832,108 +3831,75 @@ exec_stmt_change_dbowner(PLtsql_execstate *estate, PLtsql_stmt_change_dbowner *s
+		else
+			ereport(ERROR,
+				(errcode(ERRCODE_UNDEFINED_OBJECT),
+					errmsg("Table or indexed view \"%s\" does not have a full-text index or user does not have permission to perform this action.",
```
The error message mentions permissions - have we tested for cases of users that wouldn't have DROP FULLTEXT INDEX permissions

roshan0708 replied:
Actually this is a generic sql server error message which is thrown when there exist no fulltext index to be deleted from the specified table. Although according to the official docs:

> To drop a full-text index in SQL Server, you need to have ALTER permission on the table or indexed view, or be a member of the sysadmin fixed server role, or db_owner or db_ddladmin fixed database roles.

surendravishnoi replied:
we should add some test cases for permission check as well, what level of permissions apply at schema, table level etc.

Sairakan reviewed (approved):
Suggestions to clean up/clarify the code, but otherwise looks good to me. I do have the question about permission checking, but that is lower priority.

roshan0708 committed:
fixed the index creation and deletion logic

Sairakan commented on contrib/babelfishpg_tsql/src/pltsql_utils.c:
```diff
@@ -1791,3 +1792,125 @@ List
+            HeapTuple       indexTuple = SearchSysCache1(INDEXRELID, ObjectIdGetDatum(indexOid));
+            Form_pg_index   indexForm = (Form_pg_index) GETSTRUCT(indexTuple);
+
+            if (indexForm->indisunique && indexForm->indrelid == relid)
```
per https://github.com/MicrosoftDocs/sql-docs/blob/live/docs/t-sql/statements/create-fulltext-index-transact-sql.md#key-index-index_name, in addition to being unique, the index, must also be non-nullable and single-key. I am not sure if the pg_index tuple has metadata for all of these conditions, but if it does we might as well include them in this check as well.

roshan0708 committed:
fix pg_dump build fails and add unsupported stub
for ALTER FULLTEXT INDEX statements

surendravishnoi commented on contrib/babelfishpg_tsql/src/pl_exec-2.c:
```diff
@@ -52,13 +52,15 @@ static int	exec_stmt_usedb(PLtsql_execstate *estate, PLtsql_stmt_usedb *stmt);
 static int	exec_stmt_usedb_explain(PLtsql_execstate *estate, PLtsql_stmt_usedb *stmt, bool shouldRestoreDb);
 static int	exec_stmt_grantdb(PLtsql_execstate *estate, PLtsql_stmt_grantdb *stmt);
 static int	exec_stmt_grantschema(PLtsql_execstate *estate, PLtsql_stmt_grantschema *stmt);
+static int  exec_stmt_fulltextindex(PLtsql_execstate *estate, PLtsql_stmt_fulltextindex *stmt);
```
indentation

surendravishnoi commented on contrib/babelfishpg_tsql/src/pl_exec-2.c:
```diff
@@ -3817,3 +3834,86 @@ exec_stmt_change_dbowner(PLtsql_execstate *estate, PLtsql_stmt_change_dbowner *s
+	table_name = stmt->table_name;
+	is_create = stmt->is_create;
+
+	if (schema_name == NULL || strlen(schema_name) == 0 || *schema_name == '\0')
```
we are checking for NULL schema here but not in gen_schema_name_for_fulltext_index? Either put an assert that schema name should not be NULL or add checks.

surendravishnoi commented on contrib/babelfishpg_tsql/src/pltsql_utils.c:
```diff
@@ -1791,3 +1792,125 @@ List
+ * Get the fulltext index name of a relation specified by OID
+ */
+char *
+get_fulltext_index_name(Oid relid, const char *table_name)
```
indentation is messed up at number of places

surendravishnoi commented on contrib/babelfishpg_tsql/src/tsqlIface.cpp:
```diff
@@ -6580,6 +6597,102 @@ post_process_alter_table(TSqlParser::Alter_tableContext *ctx, PLtsql_stmt_execsq
+        // No "." character found, set first to the entire string
+        table_name = table_info;
+    }
+    return std::make_pair(downcase_truncate_identifier(table_name.c_str(), table_name.length(), true),
```
we should add test cases with mixed case schema, table and index name

surendravishnoi commented on contrib/babelfishpg_tsql/src/tsqlIface.cpp:
```diff
@@ -6580,6 +6597,102 @@ post_process_alter_table(TSqlParser::Alter_tableContext *ctx, PLtsql_stmt_execsq
+                           downcase_truncate_identifier(schema_name.c_str(), schema_name.length(), true));
+}
+
+PLtsql_stmt* makeCreateFulltextIndexStmt(TSqlParser::Create_fulltext_indexContext *ctx)
```
indentation

surendravishnoi commented on contrib/babelfishpg_tsql/src/pl_exec-2.c:
```diff
@@ -3817,3 +3834,86 @@ exec_stmt_change_dbowner(PLtsql_execstate *estate, PLtsql_stmt_change_dbowner *s
+		else
+			ereport(ERROR,
+				(errcode(ERRCODE_WRONG_OBJECT_TYPE),
+					errmsg("'\"%s\"' is not a valid index to enforce a full-text search key. A full-text search key must be a unique, non-nullable, single-column index which is not offline, is not defined on a non-deterministic or imprecise nonpersisted computed column, does not have a filter, and has maximum size of 900 bytes. Choose another index for the full-text key.",
```
there is no test for this, please make sure to add test case for each possible error case

roshan0708 committed:
added permission based tests, did file restructuring
, added backdoor check for CONTAINS(), updated the
fulltext GUC, updated fts-contains tests and output files,
modified and updated unsupported error messages.

roshan0708 committed:
Merge remote-tracking branch 'upstream/BABEL_3_X_DEV' into JIRA-BABEL-4383

roshan0708 committed:
Fixed the gin index issue arising because of
NULLS FIRST/LAST option

roshan0708 committed:
Fix fulltext index test files

roshan0708 committed:
Merge remote-tracking branch 'upstream/BABEL_3_X_DEV' into JIRA-BABEL-4383

roshan0708 committed:
minor code cleanup

Commit:
Create new superuser-only GUC for full text parser enablement

Signed-off-by: Jason Teng <jasonten@amazon.com>

Commit:
Update test files

Signed-off-by: Jason Teng <jasonten@amazon.com>

Commit:
Fix tests

Signed-off-by: Jason Teng <jasonten@amazon.com>

Commit:
Fix tests

Signed-off-by: Jason Teng <jasonten@amazon.com>

Commit:
Fix tests

Signed-off-by: Jason Teng <jasonten@amazon.com>

Commit:
Fix tests for pgdump

Signed-off-by: Jason Teng <jasonten@amazon.com>

Commit:
Fix tests

Signed-off-by: Jason Teng <jasonten@amazon.com>

Commit:
fix tests

Signed-off-by: Jason Teng <jasonten@amazon.com>

forestkeeper commented on contrib/babelfishpg_tsql/src/fts.c:
```diff
@@ -13,6 +14,13 @@ babelfish_fts_rewrite(PG_FUNCTION_ARGS)
+    {
+        ereport(ERROR,
+                (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+                errmsg("Full Text Search is not yet supported.")));
```
Are we sure about this error msg ? It's not supported ?

Sairakan replied:
This is a fail-safe, as we are waiting for appsec review. The idea is that until the parser is approved by appsec, it will be controlled by a superuser-only GUC. Customers who wish to experiment with full text search in red panda can submit a request for CP to override the GUC to allow them to use full text search features. We will remove this GUC in the next release after Red Panda assuming that it works well.

forestkeeper commented on contrib/babelfishpg_tsql/src/pl_exec-2.c:
```diff
@@ -3822,3 +3825,101 @@ exec_stmt_change_dbowner(PLtsql_execstate *estate, PLtsql_stmt_change_dbowner *s
+	/* The above query will be
+	 * executed using ProcessUtility()
+	 */
+	exec_utility_cmd_helper(query_str);
```
I'm a litte confused about this part, in some other impl, we do construct a structure, not a query string.

Sairakan replied:
All other cases of calling this function that I found in the code just use a StringInfoData and call the function on the char* contained within the struct, so I don't think this is that different.

forestkeeper replied:
would you mind refer to : transformAlterTableStmt , and transformSelectIntoStmt

forestkeeper commented on contrib/babelfishpg_tsql/src/pltsql_utils.c:
```diff
@@ -1791,3 +1795,204 @@ List
+	{
+		char *col_name = (char *) list_nth(column_name, i);
+		// Add column name
+		appendStringInfo(&query, "to_tsvector('fts_contains_simple', \"%s\")", col_name);
```
It's not a good idea to use string to pass a internal real cmd.

Sairakan replied:
Agreed, in the next release I plan to rework this to just generate the correct parse tree within gram-tsql-epilogue.y.c

forestkeeper commented on contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-epilogue.y.c:
```diff
@@ -1887,6 +1887,10 @@ tsql_index_nulls_order(List *indexParams)
 			continue;
 
+		/* GIN indexes don't support NULLS FIRST/LAST options */
+		if (strcmp(accessMethod, "gin") == 0)
```
Do we need to error out in here ?

Sairakan replied:
This is in order to make this feature work with Xiaohui's PR in https://github.com/babelfish-for-postgresql/babelfish_extensions/pull/1990 - in PG, GIN indexes don't support the NULL ordering which was implemented in that PR to apply to all index creation.
